### PR TITLE
clang/AMDGPU: Set amdgpu-max-num-workgroups to disable Y/Z by default

### PR DESCRIPTION
**Pull request by arsenm**
Only OpenCL supports 2d and 3d dispatches, the other languages Y and Z
dimensions are always 1. Some of the generated OpenMP functions don't
seem to get the correct attributes. The kernels do, but the callable
__omp_offloading functions are missing them for some reason.

### CODE REVIEW

**Comment by arsenm** [comment]
* **#119009** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/119009?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/119009?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: clang/include/clang/Basic/LangOptions.h [hunk 1]
```diff
@@ -706,6 +706,10 @@ class LangOptions : public LangOptionsBase {
     return OpenCL || CUDA;
   }
 
+  /// Return true if the dispatch size for an offload language only uses one
+  /// dimension.
+  bool gridSizeIsOneDimension() const { return CUDA || HIP || OpenMP; }
```
Comment (shiltian) [inline clang/include/clang/Basic/LangOptions.h:711]: CUDA and HIP don't support 3d grid? Did I misunderstand?

**Comment by shiltian** [comment]
> The kernels do, but the callable __omp_offloading functions are missing them for some reason.

The OpenMP GPU CodeGen only calls it on the target entry point.

**Comment by arsenm** [comment]
> The OpenMP GPU CodeGen only calls it on the target entry point.

It should be calling in on every single function

File: clang/include/clang/Basic/LangOptions.h (hunk 1 above)
Comment (shiltian) [reply clang/include/clang/Basic/LangOptions.h:711]: https://docs.nvidia.com/cuda/cuda-c-programming-guide/#execution-configuration Isn't it a `dim3`?

**Comment by arsenm** [comment]
So this is wrong. It was correct for openmp. Is there a language version we check can where 3d grids were introduced?

**Comment by shiltian** [comment]
> So this is wrong. It was correct for openmp.

Even OpenMP is discussing to introduce multi-dim support, though in LLVM I already implemented it as an extension. :-)

**Pull request closed without merging**